kotlinx-coroutines-core/common/src/sync/Semaphore.kt

```diff
@@ -140,74 +142,179 @@ private class SemaphoreImpl(private val permits: Int, acquiredPermits: Int) : Se
+        while (true) {
+            // Decrement the number of available permits.
+            val p = _availablePermits.getAndDecrement()
+            // Is the number of available permits greater
+            // than the maximal one due to an incorrect
+            // `release()` call without a preceding `acquire()`?
+            if (p > permits) continue
+            // The number of permits is correct, return it.
+            return p
         }
     }
 
     override fun release() {
         while (true) {
-            val p = _availablePermits.getAndUpdate { cur ->
-                check(cur < permits) { "The number of released permits cannot be greater than $permits" }
-                cur + 1
+            // Increment the number of available permits.
+            val p = _availablePermits.getAndIncrement()
+            // Is this `release` call correct and does not
+            // exceed the maximal number of permits?
+            if (p >= permits) {
+                // Revert the number of available permits
+                // back to the correct one and fail with error.
+                coerceAvailablePermitsAtMaximum()
+                error("The number of released permits cannot be greater than $permits")
             }
+            // Is there a waiter that should be resumed?
             if (p >= 0) return
+            // Try to resume the first waiter, and
+            // restart the operation if either this
+            // first waiter is cancelled or
+            // due to `SYNC` resumption mode.
             if (tryResumeNextFromQueue()) return
         }
     }
 
+    /**
+     * Changes the number of available permits to
+     * [permits] if it became greater due to an
+     * incorrect [release] call.
+     */
+    private fun coerceAvailablePermitsAtMaximum() {
+        while (true) {
+            val cur = _availablePermits.value
+            if (cur <= permits) break
+            if (_availablePermits.compareAndSet(cur, permits)) break
+        }
+    }
+
     /**
      * Returns `false` if the received permit cannot be used and the calling operation should restart.
      */
-    private fun addAcquireToQueue(cont: CancellableContinuation<Unit>): Boolean {
+    private fun addAcquireToQueue(waiter: Any): Boolean {
         val curTail = this.tail.value
         val enqIdx = enqIdx.getAndIncrement()
         val segment = this.tail.findSegmentAndMoveForward(id = enqIdx / SEGMENT_SIZE, startFrom = curTail,
             createNewSegment = ::createSegment).segment // cannot be closed
         val i = (enqIdx % SEGMENT_SIZE).toInt()
         // the regular (fast) path -- if the cell is empty, try to install continuation
-        if (segment.cas(i, null, cont)) { // installed continuation successfully
-            cont.invokeOnCancellation(CancelSemaphoreAcquisitionHandler(segment, i).asHandler)
+        if (segment.cas(i, null, waiter)) { // installed continuation successfully
+            when (waiter) {
```

Comment:
I think that the code would be much cleaner if we introduced a private sealed interface to represent the possible types of waiters. This would cost us one additional application per slow-path acquire, but get rid of `Any`, impossible `error`, etc.

Reply:
I agree with you, but your suggestion requires either `CancellableContinuation` to implement this sealed interface or adding unsafe casts at every `CancellableContinuation` usage (in case only`CancellableContinuationImpl` implements the interface). I do not see a good solution, so I would keep the current code.

Reply:
Dima, thanks, it's indeed worth doing. I'll try to do it separately in a branch with new channels, so the disturbance will be minimal. I have a prototype for selects only and it looks much more readable

Reply:
The idea was to do something like this (pseudocode):
```kotlin
private sealed interface ResumableComputation {
  fun resume()
  fun disposeOnCancellation(handle: DisposableHandle)
}

private class CancellableContinuationAsResumable(
  val continuation: CancellableContinuation<Unit>
): ResumableComputation {
  ...
}
```
So, `CancellableContinuation` would not itself implement anything (though maybe this interface makes sense in general and it should), it would be wrapped in an adapter. The cost is one more allocation on the slow path.